src/main/java/build/buf/protovalidate/internal/errors/ConstraintViolation.java

```diff
@@ -0,0 +1,268 @@
+// Copyright 2023-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buf.protovalidate.internal.errors;
+
+import build.buf.protovalidate.Violation;
+import build.buf.protovalidate.internal.evaluator.Value;
+import build.buf.validate.FieldPath;
+import build.buf.validate.FieldPathElement;
+import com.google.protobuf.Descriptors;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Deque;
+import java.util.List;
+import java.util.Objects;
+import javax.annotation.Nullable;
+
+/**
+ * {@link ConstraintViolation} contains all of the collected information about an individual
+ * constraint violation.
+ */
+public class ConstraintViolation implements Violation {
```

Comment:
Yep, same feelings here. I don't love it exactly, but actually it seems like a lot of Java software winds up doing basically this sort of thing, since there's really not much flexibility w.r.t. visibility.